core/amber/src/main/resources/python_udf/server/udf_server.py

```diff
@@ -152,9 +120,11 @@ def do_action(self, context, action: Action):
 
         else:
             raise ValueError("Unknown action {!r}".format(action.type))
+        yield self._response('success')
 
     def _delayed_shutdown(self):
         """Shut down after a delay."""
+        sleep(0.1)
```

Comment:
this sleep is new. Please comment why it is needed. Also, in general sleep() blocks a thread. Can we use something which does sleep without blocking? Something like the python equivalent of await Task.delay(x)

Reply:
This is only invoked, exactly once per worker, during Python pyarrow server shutdown process, which is invoked by the client's command `terminate`. It adds a short delay (100ms) so that the pyarrow client on the java side can use the time to close the connection peacefully.


This will be run a separate thread thus it is not blocking. We could have used `asyncio.sleep()` to make it non busy waiting, but there is no IO here and the waiting time is extremely short (100ms), thus I don't see any advantage here to optimize.

Reply:
I think invoking it in a separate thread will still block one of the cores for 100ms. But I think you mean that the actor thread can run on other cores. This is fine as the blocking time is very short.